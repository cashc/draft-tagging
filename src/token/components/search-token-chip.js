import { h } from 'react-hyperscript-helpers';
import styled from 'styled-components';
import { connect } from 'react-redux';

import { getTokensForWordsSorted } from '../selectors';

const SearchTokenChipContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  flex-direction: column;
  padding: 5px 0 18px 31px;
  background-color: ${(props) => props.theme.backgroundPrimary};
`;

const HelpTextDiv = styled.div`
  margin-bottom: 4px;
  font-size: 13px;
  letter-spacing: 0.1px;
  text-align: left;
  color: ${(props) => props.theme.foregroundPrimary};
`;

const WordRowDiv = styled.div`
  display: flex;
  align-items: center;
  margin: 6px 0;
`;

const WordDiv = styled.div`
  margin-right: 4px;
  font-size: 13px;
  font-weight: 600;
  line-height: 1.54;
  letter-spacing: 0.1px;
  text-align: left;
  color: ${(props) => props.theme.foregroundPrimary};
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
`;

const Chip = styled.div`
  display: flex;
  align-items: center;
  min-height: 22px;
  margin: 0 3px;
  padding-right: 2px;
  color: ${(props) => props.theme.foregroundLink};
  border-radius: 12px;
  box-shadow: 1px 1px 10px ${({ theme }) => theme.boxShadow};
  background-color: ${(props) => props.theme.backgroundDisabledInput};
  cursor: pointer;

  &:hover {
    text-shadow: 0 0 1px ${(props) => props.theme.infoBackground};
  }
`;

const Category = styled.span`
  padding: 0 1px 0 6px;
`;

const DropdownTriangle = styled.div`
  margin: 0px 3px -6px 2px;
  width: 0;
  height: 0;
  border: 4px solid transparent;
  border-top: 4px solid ${(props) => props.theme.foregroundLink};
`;

export const QueryHitsDiv = styled.div`
  font-size: 13px;
  letter-spacing: 0.1px;
  text-align: left;
  color: ${(props) => props.theme.foregroundSecondary};
  margin-top: 5px;
`;

export const SearchTokenChip = ({
  tokensForWords = [],
  chooseTokenCategory = () => {},
}) => {
  const tokensForWordsRows = tokensForWords.map((tokensForWord) => {
    const word = tokensForWord[0].token;
    const tokens = tokensForWord.map((token) =>
      TokenChip(token, { chooseTokenCategory }),
    );
    return h(WordRowDiv, [h(WordDiv, `${word}:`), ...tokens]);
  });

  return h(SearchTokenChipContainer, [
    h(HelpTextDiv, 'Highlight a selection you want to tag and then click on it to choose a tag.'),
    ...tokensForWordsRows,
  ]);
};

const TokenChip = (token, { chooseTokenCategory }) => {
  const { categoryTitle } = token;
  return h(
    Chip,
    {
      onClick: (event) => chooseTokenCategory(event, token),
    },
    [h(Category, categoryTitle), h(DropdownTriangle)],
  );
};

const mapStateToProps = (state) => ({
  tokensForWords: getTokensForWordsSorted(state),
});

export const mapDispatchToProps = () => ({
  chooseTokenCategory: (event, selectedToken) => {
    console.log('HIT', selectedToken);
  },
});

export const SearchTokenChipConn = connect(mapStateToProps, mapDispatchToProps)(
  SearchTokenChip,
);
