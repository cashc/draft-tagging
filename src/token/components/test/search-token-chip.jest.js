import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { h } from 'react-hyperscript-helpers';

import { SearchTokenChip, QueryHitsDiv } from '../search-token-chip';

describe('SearchTokenChip', () => {
  context('when no results', () => {
    it('should not render', () => {
      const resultsExist = false;
      const wrapper = shallow(h(SearchTokenChip, { resultsExist }));
      expect(toJson(wrapper)).toBe('');
    });
  });
  context('when no tokens but results exist', () => {
    it('should render without category', () => {
      const resultsExist = true;
      const tokensForWords = [];
      const wrapper = shallow(
        h(SearchTokenChip, { resultsExist, tokensForWords }),
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  context('when tokens grouped by word and results exist', () => {
    it('category is rendered', () => {
      const resultsExist = true;
      const tokensForWords = [
        [{ token: 'Eric', categoryTitle: 'Name' }],
        [
          { token: 'Trove', categoryTitle: 'Company Name' },
          { token: 'Trove', categoryTitle: 'Domain' },
        ],
      ];
      const wrapper = shallow(
        h(SearchTokenChip, { resultsExist, tokensForWords }),
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  context('when only 1 result', () => {
    it('should render singular verbiage', () => {
      const resultsExist = true;
      const resultsLength = 1;
      const wrapper = shallow(
        h(SearchTokenChip, { resultsExist, resultsLength }),
      );
      expect(
        wrapper
          .find(QueryHitsDiv)
          .dive()
          .text(),
      ).toBe('1 result');
    });
  });

  context('when multiple results', () => {
    it('should render plural verbiage', () => {
      const resultsExist = true;
      const resultsLength = 42;
      const wrapper = shallow(
        h(SearchTokenChip, { resultsExist, resultsLength }),
      );
      expect(
        wrapper
          .find(QueryHitsDiv)
          .dive()
          .text(),
      ).toBe('42 results');
    });
  });
});
