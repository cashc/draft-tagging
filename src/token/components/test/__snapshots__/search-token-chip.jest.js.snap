// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchTokenChip when no tokens but results exist should render without category 1`] = `
<search-token-chip__SearchTokenChipContainer>
  <search-token-chip__QueryHitsDiv>
    0 results
  </search-token-chip__QueryHitsDiv>
</search-token-chip__SearchTokenChipContainer>
`;

exports[`SearchTokenChip when tokens grouped by word and results exist category is rendered 1`] = `
<search-token-chip__SearchTokenChipContainer>
  <search-token-chip__HelpTextDiv>
    Refine your search:
  </search-token-chip__HelpTextDiv>
  <search-token-chip__WordRowDiv>
    <search-token-chip__WordDiv>
      Eric:
    </search-token-chip__WordDiv>
    <search-token-chip__Chip
      onClick={[Function]}
    >
      <search-token-chip__Category>
        Name
      </search-token-chip__Category>
      <search-token-chip__DropdownTriangle />
    </search-token-chip__Chip>
  </search-token-chip__WordRowDiv>
  <search-token-chip__WordRowDiv>
    <search-token-chip__WordDiv>
      Trove:
    </search-token-chip__WordDiv>
    <search-token-chip__Chip
      onClick={[Function]}
    >
      <search-token-chip__Category>
        Company Name
      </search-token-chip__Category>
      <search-token-chip__DropdownTriangle />
    </search-token-chip__Chip>
    <search-token-chip__Chip
      onClick={[Function]}
    >
      <search-token-chip__Category>
        Domain
      </search-token-chip__Category>
      <search-token-chip__DropdownTriangle />
    </search-token-chip__Chip>
  </search-token-chip__WordRowDiv>
  <search-token-chip__QueryHitsDiv>
    0 results
  </search-token-chip__QueryHitsDiv>
</search-token-chip__SearchTokenChipContainer>
`;
